tests/drivers/pwm/pwm_gpio_loopback/src/main.c

```diff
@@ -0,0 +1,303 @@
+/*
+ * Copyright (c) 2024 Espressif Systems (Shanghai) Co., Ltd.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief Generate PWM signals in different configurations and use a GPIO
+ * input pin to check the programmed timing. This test uses the systimer as
+ * benchmark, so it assumes the system tick is verified and precise.
+ */
+
+#include <zephyr/device.h>
+#include <inttypes.h>
+#include <zephyr/drivers/pwm.h>
+#include <zephyr/kernel.h>
+#include <zephyr/ztest.h>
+#include <zephyr/drivers/gpio.h>
+#include <stdlib.h>
+#include <math.h>
+
+#if DT_NODE_HAS_STATUS_OKAY(DT_ALIAS(pwm_0))
+#define PWM_DEV_NODE DT_ALIAS(pwm_0)
+#elif DT_NODE_HAS_STATUS_OKAY(DT_ALIAS(pwm_1))
+#define PWM_DEV_NODE DT_ALIAS(pwm_1)
+#elif DT_NODE_HAS_STATUS_OKAY(DT_ALIAS(pwm_2))
+#define PWM_DEV_NODE DT_ALIAS(pwm_2)
+#elif DT_NODE_HAS_STATUS_OKAY(DT_ALIAS(pwm_3))
+#define PWM_DEV_NODE DT_ALIAS(pwm_3)
+#else
+#error "Please define a PWM device"
+#endif
+
+#define DT_RESOURCES DT_INST(0, test_pwm_gpio_loopback)
+
+#if DT_NODE_HAS_STATUS_OKAY(DT_RESOURCES)
+#define GPIO_HDL DT_GPIO_CTLR(DT_RESOURCES, in_gpios)
+#define GPIO_DEV DEVICE_DT_GET(GPIO_HDL)
+#define GPIO_PIN DT_GPIO_PIN(DT_RESOURCES, in_gpios)
+#define GPIO_FLG DT_GPIO_FLAGS(DT_RESOURCES, in_gpios)
+#else
+#error Invalid device tree config for GPIO input pin
+#endif
+
+/* Variables for timing measurement */
+static struct test_context {
+	const struct device *pwm_dev;
+	uint32_t last_edge_time;
+	uint32_t high_time;
+	uint32_t low_time;
+	bool sampling_done;
+	uint8_t skip_cnt;
+} ctx;
+
+#if defined(CONFIG_SOC_SERIES_ESP32S2) || defined(CONFIG_SOC_SERIES_ESP32S3) ||                    \
+	defined(CONFIG_SOC_SERIES_ESP32C3)
+#define DEFAULT_PERIOD_CYCLE 16200
+#define DEFAULT_PULSE_CYCLE  8100
+#define DEFAULT_PERIOD_NSEC  160000
+#define DEFAULT_PULSE_NSEC   40000
+#else
+#define DEFAULT_PERIOD_CYCLE 64000
+#define DEFAULT_PULSE_CYCLE  32000
+#define DEFAULT_PERIOD_NSEC  2000000
+#define DEFAULT_PULSE_NSEC   500000
+#endif
```

Comment:
Board specific or SoC specific configs should be moved to Kconfig, and defined using snippets or board overlays, see https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/drivers/comparator/gpio_loopback/snippets and https://github.com/zephyrproject-rtos/zephyr/blob/d789f339efb78a136a10c74efdeeb963e085e104/tests/drivers/comparator/gpio_loopback/testcase.yaml#L14-L20

This allows support for new boards/socs to be added without altering the source code